### PR TITLE
JAMES-2132 MDN report generation library

### PR DESCRIPTION
**Pull request by chibenwa**
- Move the code as a top level **james-project** maven module (think library for parsing MDN)
 - Upgrade the code (builders, unit tests everywhere, preconditions, etc...)
 - Conform to the RFC-8098 . Here is the defect list:

    Typo Recepients -> Recipients
    Most fields are optional
    Most field values is compulsory
    Additional Disposition types
    Arbitrary Disposition modifiers
    Missing MDN-Gateway field
    Missing Errors fields
    Missing user defined fields
    Text can be multiline
    Address types needs to be flexible

I appologize, most of the Bean related boilerstuff thing are in the last commit. I did not rebase to avoid conflicts, and wanted to preserve the RFC compliance steps...

### CODE REVIEW

File: mdn/pom.xml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        <groupId>org.apache.james</groupId>
+        <version>3.1.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>james-mdn</artifactId>
+
+
+    <name>Apache James :: MDN</name>
+    <description>Provides parser and representations for MDN messages</description>
```
Comment (mbaechler) [inline mdn/pom.xml]: description doesn't explain why you need mailet-base

File: mdn/src/main/java/org/apache/james/mdn/action/mode/DispositionActionMode.java [hunk 2]
```diff
@@ -19,11 +19,30 @@
...
  * Interface <code>DispositionActionMode</code> marks a type encapsulating
  * disposition action mode information as defined by RFC 2298.
  */
-public interface DispositionActionMode
-{
+public enum DispositionActionMode {
+    Manual("manual-action"),
+    Automatic("automatic-action");
+
+    public static Optional<DispositionActionMode> fromString(String value) {
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/action/mode/DispositionActionMode.java]: no tests ?

File: mdn/src/main/java/org/apache/james/mdn/MDNFactory.java [hunk 3]
```diff
@@ -28,39 +28,19 @@
...
      */
-    static public MimeMultipartReport create(String humanText,
-            String reporting_UA_name,
-            String reporting_UA_product,
-            String original_recipient,
-            String final_recipient,
-            String original_message_id,
-            Disposition disposition) throws MessagingException
-    {
+    public static MimeMultipartReport create(String humanText,
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/MDNFactory.java]: MDNFactory.create doesn't return an MDN ?

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            return this;
+        }
+
+        public Builder dispositionField(Disposition dispositionField) {
+            this.dispositionField = Optional.of(dispositionField);
+            return this;
+        }
+
+        public MDNReport build() {
+            Preconditions.checkState(reportingUserAgentField.isPresent());
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: what is the point of using optional if it's not optional ?

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        return originalMessageIdField;
+    }
+
+    public Disposition getDispositionField() {
+        return dispositionField;
+    }
+
+    public String formattedValue() {
+        return reportingUserAgentField.formattedValue() + LINE_END
+            + originalRecipientField.map(value -> value.formattedValue() + "\r\n").orElse("")
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: LINE_END ?

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    public OriginalMessageId getOriginalMessageIdField() {
+        return originalMessageIdField;
+    }
+
+    public Disposition getDispositionField() {
+        return dispositionField;
+    }
+
+    public String formattedValue() {
+        return reportingUserAgentField.formattedValue() + LINE_END
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: Joiner(LINE_END) ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/RejectAction.java [hunk 7]
```diff
@@ -108,17 +114,26 @@ public void execute(ActionReject anAction, Mail aMail, ActionContext context) th
...
         String original_message_id = aMail.getMessage().getMessageID();
 
-        DispositionModifier modifiers[] = {DispositionModifier.Error};
-        Disposition disposition = new Disposition(DispositionActionMode.Automatic,
-                DispositionSendingMode.Automatic, DispositionType.Deleted, modifiers);
-
-        MimeMultipart multiPart = MDNFactory.create(humanText.toString(),
-                reporting_UA_name, reporting_UA_product, original_recipient,
-                final_recipient, original_message_id, disposition);
+        MimeMultipart multiPart = MDNFactory.create(
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/RejectAction.java]: this call is too complex, the API is probably not good enough judging on simplicity criteria

File: mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java [hunk 8]
```diff
@@ -537,4 +538,66 @@ public void generateMDNReportThrowOnNullDisposition() {
...
+            .gatewayField(new Gateway("host.com"))
+            .finalRecipientField(new FinalRecipient(Optional.of("final_recipient")))
+            .originalRecipientField(new OriginalRecipient("originalRecipient"))
+            .originalMessageIdField(new OriginalMessageId(Optional.of("original_message_id")))
+            .dispositionField(disposition)
+            .build()
+            .formattedValue();
+
+        assertThat(report)
+            .isEqualTo("Reporting-UA: UA_name; UA_product\r\n" +
```
Comment (mbaechler) [inline mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java]: usage of isEqualTo doesn't make sense : you want to make sure Gateway is present with a given content and maybe where it's located but not more than that.

File: mdn/src/main/java/org/apache/james/mdn/fields/OriginalRecipient.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mdn.fields;
+
+import org.apache.james.mdn.Constants;
+
+public class OriginalRecipient implements Field{
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/fields/OriginalRecipient.java]: missing space

File: mdn/src/main/java/org/apache/james/mdn/MDNFactory.java [hunk 10]
```diff
@@ -72,6 +72,28 @@ static public MimeMultipartReport create(String humanText,
...
+        String mdnReport = generateMDNReport(reporting_UA_name,
+            reporting_UA_product,
+            original_recipient,
+            final_recipient,
+            original_message_id,
+            disposition);
+
+        MimeBodyPart mdnPart = new MimeBodyPart();
+        mdnPart.setContent(mdnReport, "message/disposition-notification");
+        multiPart.addBodyPart(mdnPart);
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/MDNFactory.java]: maybe split into different methods instead of comments part 1 / part 2 / part 3

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java [hunk 11]
```diff
@@ -74,7 +86,7 @@ public MDNReport build() {
             Preconditions.checkState(dispositionField.isPresent());
 
             return new MDNReport(reportingUserAgentField,
-                originalRecipientField,
+                gatewayField, originalRecipientField,
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: please add c/r

File: mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java [hunk 12]
```diff
@@ -330,88 +330,6 @@ public void generateMDNReportShouldFormatNoModifier() {
     }
 
     @Test
-    public void generateMDNReportShouldFormatNoModifierNullType() {
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java]: you are missing tests for null/absent cases, showing that it throws?
please don't forget that you should throw when generating the report, not when populating the data structure, else you won't be able to read bad data when you implement a parser

File: mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java [hunk 13]
```diff
@@ -58,8 +58,8 @@ public void generateMDNReportShouldFormatAutomaticActions() {
             .reportingUserAgentField(new ReportingUserAgent(
                 "UA_name",
                 Optional.of("UA_product")))
-            .finalRecipientField(new FinalRecipient("final_recipient"))
-            .originalRecipientField(new OriginalRecipient("originalRecipient"))
+            .finalRecipientField(new FinalRecipient(Text.fromRawText("final_recipient")))
+            .originalRecipientField(new OriginalRecipient(Text.fromRawText("originalRecipient")))
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java]: fromRawString?

File: mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java (hunk 12 above)
Comment (chibenwa) [reply mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java]: Not if the lenient parser corrects the value ;-)

File: mdn/pom.xml (hunk 1 above)
Comment (chibenwa) [reply mdn/pom.xml]: I will remove dependency.

File: mdn/src/main/java/org/apache/james/mdn/action/mode/DispositionActionMode.java (hunk 2 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/action/mode/DispositionActionMode.java]: Added later. Sorry.

File: mdn/src/main/java/org/apache/james/mdn/MDNFactory.java (hunk 3 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNFactory.java]: I need to introduce a top level MDN object.

And we should be able to do:

```
mdn.asMimeMessage();
```

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java (hunk 4 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: 1. It becomes optional later in the PR
2. I really don't like null values.

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java (hunk 5 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: ... Done later... Sorry again

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/RejectAction.java (hunk 7 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/RejectAction.java]: We should also expose API calls that do nt require to create the fields. Will do.

File: mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java (hunk 8 above)
Comment (chibenwa) [reply mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java]: Here I value the fact of having explicit full MDN values, to show real word example with real world value.

If I am really asked for it, I will reduce it to a simple **contains** call, but I would prefer not to.

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java (hunk 11 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: Done later, sorry...

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: Done later, sorry...

File: mdn/src/main/java/org/apache/james/mdn/fields/Disposition.java [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    public List<DispositionModifier> getModifiers() {
+        return modifiers;
+    }
+
+    @Override
+    public String formattedValue() {
+        return "Disposition: "
+            + actionMode.map(DispositionActionMode::getValue).orElse("") + "/"
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/fields/Disposition.java]: The spec do not allow the field to be empty, you see here only refactored version.

I solve this in https://github.com/linagora/james-project/pull/951/commits/294d53b8cffed6b122d52b02d0c7e3b467dc4c52

File: mdn/src/main/java/org/apache/james/mdn/MDNFactory.java (hunk 10 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNFactory.java]: I need to extract this in a MDN object, as discussed with @mbaechler . Will Do.

File: mdn/src/main/java/org/apache/james/mdn/MDNFactory.java (hunk 3 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNFactory.java]: As explained earlier, I will introduce a MDN object.

File: mdn/src/main/java/org/apache/james/mdn/action/mode/DispositionActionMode.java (hunk 2 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/action/mode/DispositionActionMode.java]: Later on. See https://github.com/linagora/james-project/pull/951/commits/937b6950c22c8abb00b75409c0d7ed6be24ae970

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java (hunk 5 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: Explained in your earlier (duplicated) comment

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/RejectAction.java (hunk 7 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/RejectAction.java]: Explained in your earlier (duplicated) comment

File: mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java (hunk 8 above)
Comment (chibenwa) [reply mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java]: I attempted to start a discussion on this in your earlier comments.

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java [hunk 15]
```diff
@@ -170,8 +175,8 @@ public Disposition getDispositionField() {
         return gatewayField;
     }
 
-    public Optional<Error> getErrorField() {
-        return errorField;
+    public List<Error> getErrorField() {
```
Comment (quynhn) [inline mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: s/getErrorField/getErrorFields??

File: mdn/src/main/java/org/apache/james/mdn/fields/FinalRecipient.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+ ****************************************************************/
+
+package org.apache.james.mdn.fields;
+
+import java.util.Optional;
+
+import org.apache.james.mdn.Constants;
+
+public class FinalRecipient implements Field {
+    public static final String FIELD_NAME = "Final-Recepient";
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/fields/FinalRecipient.java]: IMO it is better for it to be public

File: mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java [hunk 17]
```diff
@@ -334,8 +556,14 @@ public void generateMDNReportShouldFormatNullOriginalMessageId() {
     @Test
     public void generateMDNReportThrowOnNullDisposition() {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java]: throws is shorter than  "throws Exception" while not loosing information, I prefer it.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/RejectAction.java (hunk 7 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/jsieve/RejectAction.java]: Simplified

File: mdn/src/main/java/org/apache/james/mdn/MDNFactory.java (hunk 3 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNFactory.java]: Done

File: mdn/pom.xml (hunk 1 above)
Comment (chibenwa) [reply mdn/pom.xml]: Done

File: mdn/src/main/java/org/apache/james/mdn/MDNReport.java (hunk 6 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNReport.java]: Done

File: mdn/src/main/java/org/apache/james/mdn/fields/OriginalRecipient.java (hunk 9 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/fields/OriginalRecipient.java]: DONE

File: mdn/src/main/java/org/apache/james/mdn/MDNFactory.java (hunk 10 above)
Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDNFactory.java]: Done

File: mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java (hunk 12 above)
Comment (chibenwa) [reply mdn/src/test/java/org/apache/james/mdn/MDNFactoryTest.java]: Such test are BTW implemented in https://github.com/linagora/james-project/pull/951/commits/937b6950c22c8abb00b75409c0d7ed6be24ae970

File: mdn/src/main/java/org/apache/james/mdn/MDN.java [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+    public MDNReport getReport() {
+        return report;
+    }
+
+    public MimeMultipart asMultipart() throws MessagingException {
+        MimeMultipart multipart = new MimeMultipart();
+        multipart.setSubType("report");
+        multipart.addBodyPart(computeHumanReadablePart());
+        multipart.addBodyPart(computeReportPart());
+        // Part 3: The optional third part, the original message is omitted.
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/MDN.java]: We can remove "Part 3:". The rest of the text is explicit, while you have no Part 1: nor Part 2:

Comment (chibenwa) [reply mdn/src/main/java/org/apache/james/mdn/MDN.java]: Ok

**Comment by aduprat** [comment]
I'm quite not sure that not checking if strings are empty is a good thing in Fields implementations
and also in MDN.Builder for humanReadableText.

**Comment by chibenwa** [comment]
> I'm quite not sure that not checking if strings are empty is a good thing in Fields implementations
and also in MDN.Builder for humanReadableText.

You are right, it is a must have.

**Comment by chibenwa** [comment]
Merged
